### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by elandorr** [theacodes/cmarkgfm#53]
Memory doesn't seem to ever get released.

Run `github_flavored_markdown_to_html()` without any assignments or anything else in a loop, and it will blow up until windows BSODs. (Haven't tried on Linux yet, as this is supposed to be used on win as well.)

Am I missing some crucial step, or is this a bug? I tried manually calling the GC in the loop, but that only made it slower, and didn't release memory.

**Comment by elandorr** [comment]
An example to test:
```py
import cmarkgfm
import memory_profiler

md = """
bla*bla*bla
# Header 1
## Header 2
### Header 3
#### Header 4
##### Header 5
###### Header 6
"""
md = md * 32


@profile
def test():
    for i in range(10000):
        cmarkgfm.github_flavored_markdown_to_html(md)


test()
```

Profiler so you don't have to watch live. I tried splitting the loops up into chunks, waiting between chunks, but nope. An ordinary size md document * 1000 is enough to BSOD a typical office  computer.

**Comment by SteinRobert** [comment]
Just stumbled upon this - seems like you want to release the memory again right?
Python itself does not take too much care of releasing memory itself. So I guess that's not a problem with this package itself.

A good way would probably be to chunk your workload and spawn child processes which themselves do the memory intense work (`github_flavored_markdown_to_html`). As soon as these processes have finished (and exited) their memory is released again.

There is also a good  post on Stackoverflow for this:
https://stackoverflow.com/questions/15455048/releasing-memory-in-python

**Comment by elandorr** [comment]
Thank you for your post @SteinRobert!

While python may not have the most efficient GC, I can't say I ever encountered practical issues. Wanting it to be less for lightweight-bliss sure, but never an explosion.

> I guess that's not a problem with this package itself.

It is, any other cffi wrapper I use doesn't face memory leaks. A C-friend looked over this for me in June and confirmed this one just doesn't handle memory management.

Looks like it could be quite simple to fix: https://cffi.readthedocs.io/en/latest/ref.html?highlight=gc#ffi-gc-1

Found a bunch of others dealing with it https://stackoverflow.com/questions/36294435/python-cffi-memory-management-issues, and plenty of 'correct' implementations. python-xxhash for example used to use ffi.gc, but then switched to handling it manually. But I'll leave that to whoever wants to dive into cffi properly instead of doing something hackish.

As this is not my forte and I didn't want to waste time researching cffi when I never use it directly, I instead spent an inordinate amount of time creating a multi-core multi-thread wrapper around the wrapper to keep memory in check, similar to what you suggest. To keep speeds useful parameters need to be so high though, RAM usage is still crazy in comparison to what could be next to nil, but it works.

Could've used the upstream C binary directly then. But if I remember correctly I didn't, because building on windows is generally hell and forces you to use the gigantic MS spyware.
(Just checked, cmark-gfm has no special deps and supports cross compiles with a decent compiler, mingw, fantastic! Well, too late.)

There is a problem with this approach though:
A cffi wrapper that handles memory is extremely much faster than spawning a gazillion processes. I benchmarked it; the impact of this workaround is in the 1000s of orders of magnitude even with optimized params. But the underlying cmark-gfm is very fast, so it still worked out for reasonable workloads.
I can't exaggerate how fast it is, amazing little lib and commonmark compliant. :)

Maybe cffi is not that difficult and I could've just PR'd a fix, but I tend to over-research and go down rabbitholes, and needed to get it done.

Have a nice weekend!
